### DISCUSSION TITLE
`allowed_block_types_all` and backwards compatibility

### DISCUSSION THREAD

**Discussion by scottfennell-toptal** [WordPress/gutenberg#45760] [Developer Experience]
Hello,

I use `allowed_block_types_all` on all of my projects in order to limit my users to just the block types that make sense for their site.  This is very important, otherwise the vast amount of irrelevant block types are confusing.

I am finding that when updating WordPress core, I sometimes get breakage due to, for example, needing to add support for `core/list-item` in order to support `core/list`.

I think perhaps block types should be registered with dependencies, similar to enqueueing a script and declaring that it depends on some JS library.  In this example, sure, core could add `core/list-item`, but in doing so it could declare `core/list` as a dependency, so that this would have been a backwards compatible change.

If back compat is not compelling, then can `allowed_block_types_all` be changed such that the first arg is an array of all registered block types, rather than just simply a boolean `true`?  This would be a convenient way for me to inspect the list of core block types.  As it stands now, it's not clear to me how I'm supposed to discover that list programmatically.

**Comment by cbirdsong** [comment]
We also build our themes with a block whitelist and ran into this exact issue with `core/list-item`. It would definitely make sense for something like block dependencies to exist for blocks that are highly intertwined with each other, as there isn't much reason to add/remove support for `core/column` without including `core/columns` or vice versa.

---

> This would be a convenient way for me to inspect the list of core block types. As it stands now, it's not clear to me how I'm supposed to discover that list programmatically.

The best way I've found to access this is using `wp.blocks.getBlockTypes()` inside a Javascript file enqueued with various editor dependencies.

Just the JS:
```js
wp.domReady(function () {
	console.log(wp.blocks.getBlockTypes());
});
```

As an inline script that doesn't do anything else:
```php
wp_register_script( 'editor-blocks-active', '', ["wp-blocks", "wp-dom-ready"]);
wp_enqueue_script( 'editor-blocks-active' );
wp_add_inline_script( 'editor-blocks-active', "wp.domReady(function () { console.log(wp.blocks.getBlockTypes()); });");
```

**Comment by mikeybinns** [comment]
You can get the list of all registered blocks via this method:
`WP_Block_Type_Registry::get_instance()->get_all_registered()`
This returns an array of `WP_Block_Type`, so you need to map this to the block name for use, and you must be careful to not overwrite any filtering from other hooks (e.g. from plugins) so you must only pull in the array if it's `true`.
Full example:
```php
$allowed_block_types = ( true === $allowed_block_types ) ? array_values(
    array_map(
      function ( $block ) {
        return $block->name;
      },
      WP_Block_Type_Registry::get_instance()->get_all_registered()
    )
  ) : $allowed_block_types;
```
EDIT:
You can actually just use array_keys instead of mapping the value :) 
```php
$allowed_block_types = ( true === $allowed_block_types ) ? array_keys(
  WP_Block_Type_Registry::get_instance()->get_all_registered()
) : $allowed_block_types;
```
